File: app/src/main/java/miui/statusbar/lyric/utils/ActivityUtils.java
package miui.statusbar.lyric.utils;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.widget.Toast;
import miui.statusbar.lyric.R;
import miui.statusbar.lyric.config.ApiListConfig;
import miui.statusbar.lyric.config.Config;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Objects;

import static android.content.Context.MODE_PRIVATE;

public class ActivityUtils {

    public static String getLocalVersion(Context context) {
        String localVersion = "";
        try {
            PackageManager packageManager = context.getPackageManager();
            @SuppressLint("WrongConstant")
            PackageInfo packageInfo = packageManager.getPackageInfo(context.getPackageName(), MODE_PRIVATE);
            localVersion = packageInfo.versionName;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return localVersion;
    }

    //检查更新
    public static void checkUpdate(Activity activity) {
        Handler handler = new Handler(Looper.getMainLooper(), message -> {
            String data = message.getData().getString("value");
            try {
                JSONObject jsonObject = new JSONObject(data);
                if (!getLocalVersion(activity).equals("")) {
                    if (Integer.parseInt(jsonObject.getString("tag_name").split("v")[1])
                            > getLocalVersionCode(activity)) {

                        new AlertDialog.Builder(activity)
                                .setTitle(String.format("%s [%s]", activity.getString(R.string.NewVer), jsonObject.getString("name")))
                                .setIcon(R.mipmap.ic_launcher)
                                .setMessage(jsonObject.getString("body").replace("#", ""))
                                .setPositiveButton(activity.getString(R.string.Update), (dialog, which) -> {
                                    try {
                                        Uri uri = Uri.parse(jsonObject.getJSONArray("assets").getJSONObject(0).getString("browser_download_url"));
                                        Intent intent = new Intent(
                                                Intent.ACTION_VIEW, uri);
                                        activity.startActivity(intent);
                                    } catch (JSONException e) {
                                        showToastOnLooper(activity, activity.getString(R.string.GetNewVerError) + e);
                                    }
                                }).setNegativeButton(activity.getString(R.string.Cancel), null).create().show();
                    } else {
                        Toast.makeText(activity, activity.getString(R.string.NoVerUpdate), Toast.LENGTH_LONG).show();
                    }
                    return true;
                }
            } catch (JSONException ignored) {
            }
            showToastOnLooper(activity, activity.getString(R.string.CheckUpdateError));
            return true;
        });
        new Thread(() -> {
            String value = HttpUtils.Get("https://api.github.com/repos/577fkj/MIUIStatusBarLyric/releases/latest");
            if (!value.equals("")) {
                Message message = handler.obtainMessage();
                Bundle bundle = new Bundle();
                bundle.putString("value", value);
                message.setData(bundle);
                handler.sendMessage(message);
            } else {
                showToastOnLooper(activity, activity.getString(R.string.CheckUpdateFailed));
            }
        }).start();
    }

    //清除配置
    public static void cleanConfig(Activity activity) {
        showToastOnLooper(activity, activity.getString(R.string.ResetSuccess));
        activity.getSharedPreferences("miui.statusbar.lyric_preferences", 0).edit().clear().apply();
        PackageManager packageManager = Objects.requireNonNull(activity).getPackageManager();
        packageManager.setComponentEnabledSetting(new ComponentName(activity, "miui.statusbar.lyric.launcher"), PackageManager.COMPONENT_ENABLED_STATE_ENABLED, PackageManager.DONT_KILL_APP);
        ConfigUtils.getSP(activity, "Lyric_Config").edit().clear().apply();
        ConfigUtils.getSP(activity, "AppList_Config").edit().clear().apply();
        ConfigUtils.getSP(activity, "Icon_Config").edit().clear().apply();
        activity.finishAffinity();

    }

    public static boolean isApi(PackageManager packageManager, String packName) {
        try {
            ApplicationInfo appInfo = packageManager.getApplicationInfo(packName, PackageManager.GET_META_DATA);
            if (appInfo.metaData != null) {
                return appInfo.metaData.getBoolean("XStatusBarLyric", false);
            } else {
                return false;
            }
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
            return false;
        }
    }

    // 弹出toast
    public static void showToastOnLooper(final Context context, final String message) {
        try {
            Handler handler = new Handler(Looper.getMainLooper());
            handler.post(() -> Toast.makeText(context, message, Toast.LENGTH_LONG).show());
        } catch (RuntimeException e) {
            e.printStackTrace();
        }
    }

    public static void getNotice(Activity activity) {
        Handler handler = new Handler(Looper.getMainLooper(), message -> {
            String data = message.getData().getString("value");
            try {
                JSONObject jsonObject = new JSONObject(data);
                if (jsonObject.getString("versionCode").equals(String.valueOf(getLocalVersionCode(activity)))) {
                    if (Boolean.parseBoolean(jsonObject.getString("forcibly"))) {
                        new AlertDialog.Builder(activity)
                                .setTitle(activity.getString(R.string.NewNotice))
                                .setIcon(R.mipmap.ic_launcher)
                                .setMessage(jsonObject.getString("data"))
                                .setNegativeButton(activity.getString(R.string.Done), null)
                                .create()
                                .show();
                    }
                }
                return true;
            } catch (JSONException ignored) {
            }
            showToastOnLooper(activity, activity.getString(R.string.GetNewNoticeError));
            return true;
        });
        new Thread(() -> {
            String value = HttpUtils.Get("https://app.xiaowine.cc/app/notice.json");
            if (!value.equals("")) {
                Message message = handler.obtainMessage();
                Bundle bundle = new Bundle();
                bundle.putString("value", value);
                message.setData(bundle);
                handler.sendMessage(message);
            }
        }).start();
    }

    public static int getLocalVersionCode(Context context) {
        int localVersion = 0;
        try {
            PackageManager packageManager = context.getPackageManager();
            PackageInfo packageInfo = packageManager.getPackageInfo(context.getPackageName(), 0);
            localVersion = packageInfo.versionCode;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return localVersion;
    }
}

File: app/src/main/java/miui/statusbar/lyric/utils/ActivityUtils.kt
package miui.statusbar.lyric.utils

import android.app.Activity
import android.app.AlertDialog
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.widget.Toast
import miui.statusbar.lyric.BuildConfig
import miui.statusbar.lyric.R
import miui.statusbar.lyric.utils.HttpUtils.Get
import org.json.JSONException
import org.json.JSONObject


object ActivityUtils {
    // 弹出toast
    @JvmStatic
    fun showToastOnLooper(context: Context?, message: String?) {
        try {
            val handler = Handler(Looper.getMainLooper())
            handler.post { Toast.makeText(context, message, Toast.LENGTH_LONG).show() }
        } catch (e: RuntimeException) {
            e.printStackTrace()
        }
    }

    //清除配置
    @JvmStatic
    fun cleanConfig(activity: Activity) {
        activity.getSharedPreferences("miui.statusbar.lyric_preferences", 0).edit().clear().apply()
        val packageManager: PackageManager = activity.packageManager
        packageManager.setComponentEnabledSetting(
            ComponentName(activity, "miui.statusbar.lyric.launcher"),
            PackageManager.COMPONENT_ENABLED_STATE_ENABLED,
            PackageManager.DONT_KILL_APP
        )
        for (name in arrayOf("Lyric_Config", "AppList_Config", "Icon_Config")) {
            Utils.getSP(activity, name)?.edit()?.clear()?.apply()
        }
        showToastOnLooper(activity, activity.getString(R.string.ResetSuccess))
        activity.finishAffinity()
    }

    //检查更新
    fun checkUpdate(activity: Activity) {
        val handler = Handler(Looper.getMainLooper()) { message: Message ->
            val data: String = message.data.getString("value") as String
            try {
                val jsonObject = JSONObject(data)
                if (jsonObject.getString("tag_name").split("v").toTypedArray()[1].toInt() > BuildConfig.VERSION_CODE) {
                    AlertDialog.Builder(activity).apply {
                        setTitle(
                            String.format(
                                "%s [%s]",
                                activity.getString(R.string.NewVer),
                                jsonObject.getString("name")
                            )
                        )
                        setIcon(R.mipmap.ic_launcher)
                        setMessage(jsonObject.getString("body").replace("#", ""))
                        setPositiveButton(R.string.Update) { _, _ ->
                            try {
                                val uri: Uri = Uri.parse(
                                    jsonObject.getJSONArray("assets").getJSONObject(0)
                                        .getString("browser_download_url")
                                )
                                val intent = Intent(
                                    Intent.ACTION_VIEW, uri
                                )
                                activity.startActivity(intent)
                            } catch (e: JSONException) {
                                showToastOnLooper(activity, activity.getString(R.string.GetNewVerError) + e)
                            }
                        }
                        setNegativeButton(R.string.Cancel, null)
                        show()
                    }
                } else {
                    Toast.makeText(activity, activity.getString(R.string.NoVerUpdate), Toast.LENGTH_LONG).show()
                }
            } catch (ignored: JSONException) {
            }
            showToastOnLooper(activity, activity.getString(R.string.CheckUpdateError))
            true
        }
        Thread {
            val value: String =
                HttpUtils.Get("https://api.github.com/repos/577fkj/MIUIStatusBarLyric/releases/latest")
            if (value != "") {
                handler.obtainMessage().let {
                    it.data = Bundle().apply {
                        putString("value", value)
                    }
                    handler.sendMessage(it)
                }
            } else {
                showToastOnLooper(activity, activity.getString(R.string.CheckUpdateFailed))
            }
        }.start()
    }

    fun isApi(packageManager: PackageManager, packName: String?): Boolean {
        return try {
            val appInfo = packageManager.getApplicationInfo(packName!!, PackageManager.GET_META_DATA)
            if (appInfo.metaData != null) {
                appInfo.metaData.getBoolean("XStatusBarLyric", false)
            } else {
                false
            }
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            false
        }
    }

    fun getNotice(activity: Activity) {
        val handler = Handler(Looper.getMainLooper()) { message: Message ->
            try {
                val jsonObject = JSONObject(message.data.getString("value")!!)
                if (jsonObject.getString("versionCode") == BuildConfig.VERSION_CODE.toString()) {
                    if (java.lang.Boolean.parseBoolean(jsonObject.getString("forcibly"))) {
                        AlertDialog.Builder(activity)
                            .setTitle(activity.getString(R.string.NewNotice))
                            .setIcon(R.mipmap.ic_launcher)
                            .setMessage(jsonObject.getString("data"))
                            .setNegativeButton(activity.getString(R.string.Done), null)
                            .create()
                            .show()
                    }
                }
                return@Handler true
            } catch (ignored: JSONException) {
            }
            showToastOnLooper(activity, activity.getString(R.string.GetNewNoticeError))
            false
        }
        Thread {
            val value = Get("https://app.xiaowine.cc/app/notice.json")
            if (value != "") {
                val message = handler.obtainMessage()
                val bundle = Bundle()
                bundle.putString("value", value)
                message.data = bundle
                handler.sendMessage(message)
            }
        }.start()
    }
}